### ISSUE TITLE
Unexpected heap usage and execution time

### ISSUE THREAD

**Issue by mvanbrab** [comunica/comunica#816]
#### Issue type:

- :snail: Performance issue 

____
#### Description:

Observed heap usage exceeding 1.5 gigabytes and execution times for some queries of 230+ seconds, unexpected given the small size of the datasources involved.

Observed while using Comunica in a [Walder](https://github.com/KNowledgeOnWebScale/walder) project,
however for the purpose of this issue **isolated in [this small command line test app](https://github.com/mvanbrab/comunica-test-knows)**. This test app is recommended to reproduce the issue.

Full query input (GrapQL query, JSON-LD context, datasources) and query output available in the **log files** of that test app.

An overview of the observations is available in the **OBSERVATIONS.md** file of that test app.

A summary showing results for different node versions:

![afbeelding](https://user-images.githubusercontent.com/6448245/117263455-9416b700-ae52-11eb-9791-9db073b5d435.png)

![afbeelding](https://user-images.githubusercontent.com/6448245/117263630-be687480-ae52-11eb-8aef-e78f1e0e6ca2.png)

____
#### Environment:

Output in the test app of `node ./node_modules/@comunica/actor-init-sparql/bin/query.js -v`:
```
| software            | version
| ------------------- | -------
| Comunica Init Actor | 1.21.1 
| node                | v14.16.1
| npm                 | 6.14.12
| yarn                | 1.22.5
| Operating System    | linux (Linux 4.15.0-142-generic)
```
Note: tested with multiple node versions, see description above.

**Comment by mvanbrab** [comment]
This heap snapshot observation (by @pheyvaer) made from inside the Walder application with the same queries, may point into the direction of pending unhandled rejections:
![afbeelding](https://user-images.githubusercontent.com/6448245/117268176-4badc800-ae57-11eb-9022-84951ffad1cf.png)
They take up 300+ MB at the time the snapshot was taken.

**Comment by rubensworks** [comment]
Closing this, as this should have been resolved in #826.
Feel free to comment in case this should still be an issue.
